### ISSUE TITLE
Cast with incorrect alignment

### ISSUE THREAD

**Issue by kpeeters** [opencv/opencv#24265]
### System Information

OpenCV version: 4.8.0
OS / Platform: macOS 13.5.1 on M1
Compiler: emscripten 3.1.45 (any recent llvm)

### Detailed description

Building a project that includes opencv headers with a recent llvm and `-Wcast-align` turned on produces errors of the form

```
opencv/modules/core/include/opencv2/core/mat.inl.hpp:947:17: error: cast from 'uchar *' (aka 'unsigned char *') to 'double *' increases required alignment from 1 to 8 [-Werror,-Wcast-align]
  947 |         return *(_Tp*)(data + step.p[0] * i0);
```

This may well be correct, but should then have an `std::assume_aligned` or similar to make the compiler happy.

### Steps to reproduce

Any project that uses `-Wcast-align` and includes `opencv.hpp`.

### Issue submission checklist

- [X] I report the issue, it's not a question
- [X] I checked the problem with documentation, FAQ, open issues, forum.opencv.org, Stack Overflow, etc and have not found any solution
- [X] I updated to the latest OpenCV version and the issue is still there
- [ ] There is reproducer code and related data files (videos, images, onnx, etc)

**Comment by TomaszTB** [comment]
Building OpenCV (4.X) with `-Wcast-align=strict` enabled produces these warnings ([warnings.txt.zip](https://github.com/user-attachments/files/16121892/warnings.txt.zip)). I cut out a bunch of warnings from 3rdparty files, and it's still a massive list. 

The warnings can be silenced by adding an intermediate cast to `void*`. E.g.
Change:
```
*(_Tp*)(data + step.p[0] * i0)
```
To:
```
*(_Tp*)((void*)(data + step.p[0] * i0))
```

Otherwise to do things properly it seems `memcpy()` should be used.  

It'd be nice to fix this, but it'd also requires a lot of changes. It also seems many other libraries don't worry about this. Do we think it's worth making the changes? If so, should we simply silence the warnings (using `void*`) or switch using `memcpy()`?

**Comment by mshabunin** [comment]
If I understand correctly `memcpy` can be used only when casting single values. So perhaps it wouldn't help in most cases where we need to cast `Mat::data` to specific pointer type.

I think we could add proposed fixes (cast to `void*` or `std::assume_aligned` or GCC/LLVM `__builtin_assume_aligned` etc.) to `Mat::ptr` and `Mat::at` methods and try to use them in other places instead of raw access to `Mat::data`. Maybe an extra `CV_DbgAssert` with alignment check could be added as well.

**Comment by TomaszTB** [comment]
Unfortunately `std::assume_aligned` is from C++ 20. I'm not very familiar with builtin functions, but we'd have to make sure they're only used when OpenCV is compiled using GCC/LLVM, right? Some compilers might not recognize it? 

I'd like to take a step towards fixing this issue, so as a quick check what do we think of this sort of changes in `Mat`?
```
template<typename _Tp> inline
_Tp& Mat::at(int i0, int i1)
{
    // some CV_DbgAsserts...
    return ((_Tp*)(data + step.p[0] * i0))[i1];
}
```
to
```
template<typename _Tp> inline
_Tp& Mat::at(int i0, int i1)
{
    // some CV_DbgAsserts...

    _Tp* ptr = ((_Tp*)((void*)(data + step.p[0] * i0))) + i1;  // get pointer to data (could use __builtin_assume_aligned instead of cast to void*)
    CV_DbgAssert(((uintptr_t) ptr & (alignof(_Tp) - 1)) == 0); // ensure the pointer is aligned to the alignment of requested type
    return *ptr;
}
```

**Comment by mshabunin** [comment]
Yes, I think something like this should work. We can enable flag `-Wcast-align` to check that it actually works within OpenCV build.
